docs/rfc/0001_ctrlkit.md

```diff
@@ -0,0 +1,325 @@
+2. `alias Kind GVK`
+    * `alias` a group/version/kind to a single word name, commonly the same name with type, to make the reference of GVK easier.
+3. `decl ControllerManager for Kind` block
+    * `decl` a controller manager for a specific kind, with a body as follows:
+    * `state` block, with a body:
+      * `name Kind` / `name []Kind` block defines a state's name and kind (where `[]Kind` means a list of resources), with a body:
+        * Each line is a selector. Supported selectors and the effects are summarized in the table below.
+    * `action` block, with a body:
+      * `ActionName(states...)` defines an action with its name and states it requires. The `states...` part can be empty. But if there's any state required, it must be declared before in some `state` block.
+4. Comments must begin with `//`.
+
+Selector | Allow Values | Effect
+:---: | :---: | :---:
+name | Constant value or reference of string value of `target` | Get the resource by name (in `target`' s namespace), e.g. `name=test` or `name=${target.Name}`
+labels/some-label | Constant value or reference of string value of `target` | Get the resources by listing with the labels (in `target`'s namespace), e.g. `labels/cronjob=${target.Name}`
+fields/some-field | Constant value or reference of string value of `target` | Get the resources by listing with the fields (in `target`'s namespace), e.g. `fields/.metadata.name=test`. Note only some of the fields are supported on the server-side. Additional fields must be set as indices of cache in `controller-runtime` framework before it can be used.
+owned | No value | The state getter will validate if the selected resources are owned by `target`. If not, the behavior differs for different states. If the state is a list, the objects not owned will be left out. If the state is a single object, an error will be returned.
+
+An example of a document describes the states and actions needed for implementing a controller of `CronJob`:
+
+```plain
+bind v1 k8s.io/api/core/v1
+bind batch/v1 k8s.io/api/batch/v1
+bind controller-kit-demo/v1 controller-kit-demo/api/v1
+
+alias CronJob controller-kit-demo/v1/CronJob
+alias Job batch/v1/Job
+
+// CronJobControllerManager declaras all the actions needed by the CronJobController.
+decl CronJobControllerManager for CronJob {
+    state {
+        jobs []Job {
+            labels/cronjob=${target.Name}
+            fields/.metadata.controller=${target.Name}
+            owned
+        }
+    }
+
+    action {
+        // List all active jobs, and update the status.
+        ListActiveJobsAndUpdateStatus(jobs)
+
+        // Clean up old jobs according to the history limit.
+        CleanUpOldJobsExceedsHistoryLimits(jobs)
+
+        // Run the next job if it's on time, or otherwise we should wait 
+        // until the next scheduled time.
+        RunNextScheduledJob()
+
+        // Update status of CronJob.
+        UpdateCronJobStatus()
+    }
+}
+```
+
+### **CLI Tool**
```

Comment:
Is it possible to use golang's some syntactic sugar to implement this abstraction instead of using an external CLI? 🤔

Reply:
That's a good question! Thanks!

I haven't come up with an elegant way to achieve all the features mentioned in this proposal. I had tried wrapping the actions with closures in the `galaxykube` project. But there're still some problems, e.g., 
  * Sync between actions/states and their names. And I think it's hard to resolve it with pure Go because Go doesn't provide a way to get the name of variables/functions at the compile time. Maybe we can count on the "%T" format verb if we declare a struct for each action.
  * Polyfill codes to get the states (before generics, currently, I'm thinking about using the generics to achieve this).
 
It's worth further consideration.